File: app/src/main/java/ru/org/adons/cblock/ui/adapter/CallLogAdapter.java
package ru.org.adons.cblock.ui.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import ru.org.adons.cblock.R;
import ru.org.adons.cblock.model.CallLogItem;

import static ru.org.adons.cblock.ui.adapter.ListItemDecorator.formatPhoneNumber;
import static ru.org.adons.cblock.ui.adapter.ListItemDecorator.getDescription;

/**
 * Recycler adapter for Call Log list
 */
public class CallLogAdapter extends RecyclerView.Adapter<CallLogAdapter.ViewHolder> {

    private final ArrayList<CallLogItem> items = new ArrayList<>();

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_call_log, parent, false);
        return new CallLogAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        CallLogItem item = items.get(position);
        holder.textPhone.setText(formatPhoneNumber(item.phoneNumber()));
        holder.textDesc.setText(getDescription(item.name(), item.date()));
        // disable checkbox if item already in block list
        if (item.isBlocked) {
            holder.checkBox.setEnabled(false);
            holder.checkBox.setChecked(true);
            holder.itemView.setOnClickListener(null);
            holder.checkBox.setOnClickListener(null);
        } else {
            holder.checkBox.setEnabled(true);
            // set selected if list scrolled
            if (item.isSelected) {
                holder.checkBox.setChecked(true);
            } else {
                holder.checkBox.setChecked(false);
            }
            //
            holder.checkBox.setOnClickListener(v -> setSelected(item, holder.checkBox));
            holder.itemView.setOnClickListener(v -> setSelected(item, holder.checkBox));
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void setItems(List<CallLogItem> newItems) {
        items.clear();
        items.addAll(newItems);
        notifyDataSetChanged();
    }

    public List<CallLogItem> getItems() {
        return items;
    }

    private void setSelected(CallLogItem item, CheckBox checkBox) {
        item.isSelected = !item.isSelected;
        checkBox.setChecked(item.isSelected);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.checkbox_is_selected) CheckBox checkBox;
        @BindView(R.id.text_view_phone_number) TextView textPhone;
        @BindView(R.id.text_view_desc) TextView textDesc;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

}

File: app/src/main/java/mobile/addons/cblock/ui/view/add/CallLogAdapter.kt
package mobile.addons.cblock.ui.view.add

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import kotlinx.android.synthetic.main.list_item_call_log.view.*
import mobile.addons.cblock.R
import mobile.addons.cblock.data.CallLogItem
import mobile.addons.cblock.ext.formatPhoneNumber
import mobile.addons.cblock.ext.getDescription
import java.util.*

/**
 * Recycler adapter for Call Log list
 */
class CallLogAdapter : RecyclerView.Adapter<CallLogAdapter.ViewHolder>() {

    private val items = ArrayList<CallLogItem>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.list_item_call_log, parent, false)
        return CallLogAdapter.ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = items[position]
        val view = holder.itemView
        view.textPhoneNumber.text = item.phoneNumber.formatPhoneNumber()
        view.textDescription.text = item.date.getDescription(item.name)
        // disable checkbox if item already in block list
        if (item.isBlocked) {
            view.checkBox.isEnabled = false
            view.checkBox.isChecked = true
            view.setOnClickListener(null)
            view.checkBox.setOnClickListener(null)
        } else {
            view.checkBox.isEnabled = true
            // set selected if list scrolled
            view.checkBox.isChecked = item.isSelected
            //
            view.checkBox.setOnClickListener { setSelected(item, view.checkBox) }
            view.setOnClickListener { setSelected(item, view.checkBox) }
        }
    }

    override fun getItemCount(): Int = items.size

    internal fun setItems(newItems: List<CallLogItem>) {
        items.clear()
        items.addAll(newItems)
        notifyDataSetChanged()
    }

    internal fun getItems(): List<CallLogItem> = items

    private fun setSelected(item: CallLogItem, checkBox: CheckBox) {
        item.isSelected = !item.isSelected
        checkBox.isChecked = item.isSelected
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

}
